cep-recipe-jinja.md

```diff
@@ -0,0 +1,125 @@
+
+```
+c_compiler:
+  - gcc
+c_compiler_version:
+  - "8.9"
+cxx_compiler:
+  - clang
+cxx_compiler_version:
+  - "12"
+```
+
+## The `cdt` function
+
+cdt stands for "core dependency tree" packages. These are repackaged from Centos.
+
+The function expands to the following: 
+
+- package-name-<cdt_name>-<cdt_arch>
+
+Where `cdt_name` and `cdt_arch` are loaded from the variant config. If they are undefined, they default to:
+
+- `cos6` for `cdt_name` on `x86_64` and `x86`, otherwise `cos7`
+- To the `platform::arch` for `cdt_arch`, except for `x86` where it defaults to `i686`.
+
+## The `pin_subpackage` function
+
+### Pin definition
+
+A pin has the following arguments:
+
+- `min_pin`: The lower bound of the dependency spec. This is expressed as a `x.x....` version where the `x` are filled in from the corresponding version of the package. For example, `x.x` would be `1.2` for a package version `1.2.3`. The resulting pin spec would look like `>=1.2` for the `min_pin` argument of `x.x` and a version of `1.2.3` for the package.
+- `max_pin`: This defines the upper bound and follows the same `x.x` semantics but adds `+1` to the last segment. For example, `x.x` would be `1.(2 + 1)` for a package version `1.2.3`. The resulting pin spec would look like `<1.3` for the `max_pin` argument of `x.x` and a version of `1.2.3` for the package.
+- `exact`: This is a boolean that specifies whether the pin should be exact. It defaults to `False`. If `exact` is `True`, the `min_pin` and `max_pin` are irrelevant. We pin to an `==` version and also include the build string exactly (e.g. `==1.2.3=h1234`).
+
+#### Example
+
+If we consider a package like `numpy-1.21.3-h123456_5` we could apply some pin expressions. 
+
+- `min_pin=x.x, max_pin=x.x` would result in `>=1.21,<1.22`
+- `min_pin=x.x.x,max_pin=x` would result in `>=1.21.3,<2`
+- `exact=True` would result in `==1.21.3=h123456_5`
+
+
+## The `pin_compatible` function
+
+Pin compatible will pin the dependency to the same version as "previously" resolved in the `host` or `build` environment. This is useful to ensure that the same package is used at run time as was used at build time.
+
+Example:
+
+```yaml
+requirements:
+  host:
+    - numpy
+  run:
+    - ${{ pin_compatible('numpy', exact=True) }}
+```
+
+## The `pin_subpackage` function
+
+Pin subpackage will pin the dependency to the same version as another sub-package from the recipe (or the current package itself).
+This is useful to ensure that multiple outputs from a recipe are linked together or to export the correct `run_exports` for a package.
+
+Example:
+
+```yaml
+outputs:
+  - package:
+      name: libfoo
+      version: "1.2.3"
+  - package:
+      name: foo
+      version: "1.2.3"
+    requirements:
+      run:
+        - ${{ pin_subpackage('libfoo', exact=True) }}
+```
+
+## The `cmp` function
+
+The `cmp` function is used to compare two versions. It returns `True` if the comparison is true and `False` otherwise.
```

Comment:
Should we add an example of how it can be used?

Reply:
Yes, what operators are available and what is the syntax?

Reply:
Added some examples!

Reply:
Ah so it's more like a version matcher. When I read `compare` I thought we were comparing two packages together like "is numpy's version greater than python's", which didn't make much sense to me.

But this is more like asking "Does the version of Python match this version expression?". So in that sense the name could be better expressed as:

- `version(python, "<3.8")`
- `version_match(python, "<3.8")`
- `match_version(python, "<3.8")`
- `satisfy(python, "<3.8")`

Also this assumes that `python` will be represented by a concrete version `x.y.z` and not a spec like `x.y.*` right? IOW, it's already resolved to a specific package record.

If that's the case, should we specify that this function can only return after the solve has completed?